File: benchmark/bench_shamir.c
/*
    Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
*/

/*
 * Benchmark Shamir Secret Sharing
 */

#include "bench.h"
#include "amcl/shamir.h"

#define MIN_TIME    5.0
#define MIN_ITERS   10

char *S_hex = "fab4ce512dff74bd9c71c89a14de5b877af45dca0329ee3fcb72611c0784fef3";

int n = 30;
int k = 20;

int main()
{
    int i, rc;

    int iterations;
    clock_t start;
    double elapsed;

    // Secret
    char s[SGS_SECP256K1];
    octet S = {0,sizeof(s),s};

    // Secret shares
    char x[n][SGS_SECP256K1];
    octet X[n];
    char y[n][SGS_SECP256K1];
    octet Y[n];

    for(i = 0; i < n; i++)
    {
        Y[i].max = SGS_SECP256K1;
        Y[i].len = SGS_SECP256K1;
        Y[i].val = y[i];

        X[i].max = SGS_SECP256K1;
        X[i].len = SGS_SECP256K1;
        X[i].val = x[i];
    }

    SSS_shares shares = {X, Y};

    // Additive share for conversion
    char sh[SGS_SECP256K1];
    octet SH = {0, sizeof(sh), sh};

    // Additional checks for verification
    char c[k][1 + SFS_SECP256K1];
    octet C[k];

    for(i = 0; i < k; i++)
    {
        C[i].max = 1 + SFS_SECP256K1;
        C[i].len = 1 + SFS_SECP256K1;
        C[i].val = c[i];
    }

    // Deterministic RNG for testing
    char seed[32] = {0};
    csprng RNG;
    RAND_seed(&RNG, 32, seed);

    // Load Secret from hex
    OCT_fromHex(&S, S_hex);

    /* Benchmark */
    print_system_info();

    printf("Timing info\n");
    printf("===========\n");

    // Shamir Secret Sharing
    iterations=0;
    start=clock();
    do
    {
        SSS_make_shares(k, n, &RNG, &shares, &S);
        iterations++;
        elapsed=(clock()-start)/(double)CLOCKS_PER_SEC;
    }
    while (elapsed<MIN_TIME || iterations<MIN_ITERS);

    elapsed= MILLISECOND * elapsed / iterations;
    printf("\tSSS_make_shares\t\t%8d iterations\t",iterations);
    printf("%8.2lf ms per iteration\n",elapsed);

    iterations=0;
    start=clock();
    do
    {
        SSS_recover_secret(k, &shares, &S);
        iterations++;
        elapsed=(clock()-start)/(double)CLOCKS_PER_SEC;
    }
    while (elapsed<MIN_TIME || iterations<MIN_ITERS);

    elapsed= MILLISECOND * elapsed / iterations;
    printf("\tSSS_recover_secret\t%8d iterations\t",iterations);
    printf("%8.2lf ms per iteration\n",elapsed);

    // Shamir to additive conversion
    iterations=0;
    start=clock();
    do
    {
        SSS_shamir_to_additive(k, X, Y, X+1, &SH);
        iterations++;
        elapsed=(clock()-start)/(double)CLOCKS_PER_SEC;
    }
    while (elapsed<MIN_TIME || iterations<MIN_ITERS);

    elapsed= MICROSECOND * elapsed / iterations;
    printf("\tSSS_shamir_to_additive\t%8d iterations\t",iterations);
    printf("%8.2lf us per iteration\n",elapsed);

    // Verifiable Secret Sharing
    iterations=0;
    start=clock();
    do
    {
        VSS_make_shares(k, n, &RNG, &shares, C, &S);
        iterations++;
        elapsed=(clock()-start)/(double)CLOCKS_PER_SEC;
    }
    while (elapsed<MIN_TIME || iterations<MIN_ITERS);

    elapsed= MILLISECOND * elapsed / iterations;
    printf("\tVSS_make_shares\t\t%8d iterations\t",iterations);
    printf("%8.2lf ms per iteration\n",elapsed);

    iterations=0;
    start=clock();
    do
    {
        rc = VSS_verify_shares(k, X, Y, C);
        iterations++;
        elapsed=(clock()-start)/(double)CLOCKS_PER_SEC;
    }
    while (elapsed<MIN_TIME || iterations<MIN_ITERS);

    if (rc != VSS_OK)
    {
        printf("FAILURE  VSS_verify_shares: %d\n", rc);
        exit(EXIT_FAILURE);
    }

    elapsed= MILLISECOND * elapsed / iterations;
    printf("\tVSS_verify_shares\t%8d iterations\t",iterations);
    printf("%8.2lf ms per iteration\n",elapsed);

    exit(EXIT_SUCCESS);
}

File: python/benchmark/bench_shamir.py
#!/usr/bin/env python3

"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
"""

import os
import sys
from bench import time_func

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from amcl import core_utils, shamir

seed_hex = "78d0fb6705ce77dee47d03eb5b9c5d30"

k = 5
n = 10


if __name__ == "__main__":
    # random number generator
    seed = bytes.fromhex(seed_hex)
    rng = core_utils.create_csprng(seed)

    ## Shamir Secret Sharing

    # Generate quantities for benchmark
    shares, S = shamir.make_shares(k, n, rng)
    S_rec = shamir.recover_secret(shares[:k])
    assert S == S_rec, "Error generating quantities for SSS benchmark"

    (X, _) = zip(*(shares[:k]))
    X = [x for x in X]
    X.pop(0)

    # Run benchmark
    fncall = lambda: shamir.make_shares(k, n, rng)
    time_func("make_shares      ", fncall, unit="ms")

    fncall = lambda: shamir.recover_secret(shares[:k])
    time_func("recover_secret   ", fncall, unit="ms")

    fncall = lambda: shamir.to_additive(shares[0], X)
    time_func("to_additive      ", fncall, unit="us")

    ## Verifiable Secret Sharing

    # Generate quantities for benchmark
    shares, checks, S = shamir.vss_make_shares(k, n, rng)
    rc = shamir.vss_verify_shares(shares[0], checks)

    assert rc == shamir.OK, "Error generating quantities for VSS benchmark - Share verification"

    S_rec = shamir.recover_secret(shares[:k])
    assert S == S_rec, "Error generating quantities for VSS benchmark - wrong recovered secret"

    # Run benchmark
    fncall = lambda: shamir.vss_make_shares(k, n, rng)
    time_func("vss_make_shares  ", fncall, unit="ms")

    fncall = lambda: shamir.vss_verify_shares(shares[0], checks)
    time_func("vss_verify_shares", fncall, unit="ms")
